packages/bull-vault/src/FlashBull.sol

```diff
@@ -1,40 +1,206 @@
+            uint256 squeethEthPrice = _getTwap(ethWSqueethPool, wPowerPerp, weth, TWAP, false);
+            (wSqueethToMint, ethFee) = _calcWsqueethToMintAndFee(_ethToCrab, squeethInCrab, ethInCrab, squeethEthPrice);
+            crabAmount = _calcSharesToMint(_ethToCrab.sub(ethFee), ethInCrab, IERC20(crab).totalSupply());
+
+            uint256 share;
+            if (IERC20(bullStrategy).totalSupply() == 0) {
+                share = ONE;
+            } else {
+                share = crabAmount.wdiv(IERC20(crab).balanceOf(bullStrategy).add(crabAmount));
+            }
+
+            // USDC we pay to flashswap
+            (ethToLend, usdcToBorrow) = IBullStrategy(bullStrategy).calcLeverageEthUsdc(crabAmount, share, ethInCrab, squeethInCrab, IERC20(crab).totalSupply());
+        }
+
+        // oSQTH-ETH swap
+        _exactInFlashSwap(
+            wPowerPerp,
+            weth,
+            _poolFee,
+            wSqueethToMint,
+            _minEthFromSqth,
+            uint8(FLASH_SOURCE.FLASH_DEPOSIT_CRAB),
+            abi.encodePacked(_ethToCrab)
+        );
+
+         // ETH-USDC swap
+        _exactInFlashSwap(
+            usdc,
+            weth,
+            _poolFee,
+            usdcToBorrow,
+            _minEthFromUsdc,
+            uint8(FLASH_SOURCE.FLASH_DEPOSIT_COLLATERAL),
+            abi.encodePacked(crabAmount, ethToLend)
+        );
+
+        // return excess eth to the user that was not needed for slippage
+        if (address(this).balance > 0) {
+            payable(msg.sender).sendValue(address(this).balance);
+        }
+
+        IERC20(bullStrategy).transfer(msg.sender, IERC20(bullStrategy).balanceOf(address(this)));
+    }
+
+    /**
+     * @notice uniswap flash swap callback function
+     * @dev this function will be called by flashswap callback function uniswapV3SwapCallback()
+     */
+    function _uniFlashSwap(UniFlashswapCallbackData memory _uniFlashSwapData) internal override {
+        if (FLASH_SOURCE(_uniFlashSwapData.callSource) == FLASH_SOURCE.FLASH_DEPOSIT_CRAB) {
+            FlashDepositCrabData memory data = abi.decode(_uniFlashSwapData.callData, (FlashDepositCrabData));
+
+            // convert WETH to ETH as Uniswap uses WETH
+            IWETH9(weth).withdraw(IWETH9(weth).balanceOf(address(this)));
+            ICrabStrategyV2(crab).deposit{value: data.ethToDepositInCrab}();
 
-    // struct FlashWithdrawData {}
-    // // TODO
+            // repay the squeeth flash swap
+            IERC20(wPowerPerp).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+        } else if (FLASH_SOURCE(_uniFlashSwapData.callSource) == FLASH_SOURCE.FLASH_DEPOSIT_COLLATERAL) {
+            FlashDepositCollateralData memory data =
+                abi.decode(_uniFlashSwapData.callData, (FlashDepositCollateralData));
 
-    constructor(address payable _bull) {
-        bullStrategy = BullStrategy(_bull);
+            IWETH9(weth).withdraw(IWETH9(weth).balanceOf(address(this)));
```

Comment:
The test function was working even before adding this line, which I think it shouldn't, right?
I think we are sending too much msg.value in the test function so even without unwrapping the test was passing